instrumentation/finagle/http/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finagle/Helpers.java

```diff
@@ -0,0 +1,218 @@
+  // time
+  public static class LocalCallGuard {
+
+    private final Local<Object> guard = new Local<>();
+
+    // is the current thread presently inside a call guarded by this LocalCallGuard?
+    public boolean isRecursed() {
+      return guard.apply().isDefined();
+    }
+
+    // safely guard the Function0 call;
+    // (Function0 -> Java Supplier, in our cases -- it wraps existing calls)
+    public <T> T guardedCall(Function0<T> f, T defaultVal) {
+      if (isRecursed()) {
+        return defaultVal;
+      }
+      return guard.let(null, f);
+    }
+  }
+
+  @SuppressWarnings({"ThrowSpecificExceptions", "CheckedExceptionNotThrown"})
+  public static Future<?> loopAdviceExit(MethodHandle handle, Future<?> ret) {
+    return LOOP_GUARD.guardedCall(
+        () ->
+            Contexts.local()
+                .let(
+                    ContextRef.INSTANCE.getKey(),
+                    // this works bc at this point in the server evaluation, the netty
+                    // instrumentation has already gone to work and assigned the context to the
+                    // local thread;
+                    //
+                    // this works specifically in finagle's netty stack bc at this point the loop()
+                    // method is running on a netty thread with the necessary access to the
+                    // java-native ThreadLocal where the Context is stored
+                    Context.current(),
+                    () -> {
+                      try {
+                        // all access to Context.current() from this point forward should now
+                        // succeed as expected
+                        return (Future<?>) handle.invoke();
+                      } catch (Throwable e) {
+                        throw new RuntimeException(e);
+                      }
+                    }),
+        ret);
+  }
+
+  @SuppressWarnings("ThrowSpecificExceptions")
+  public static Future<?> callWrite(MethodHandle handle, Future<?> ret) {
+    return WRITE_GUARD.guardedCall(
+        () -> {
+          Option<Context> ref = Contexts.local().get(ContextRef.INSTANCE.getKey());
+          Callable<Future<?>> call =
+              () -> {
+                try {
+                  return (Future<?>) handle.invoke();
+                } catch (Exception e) {
+                  // don't wrap needlessly
+                  throw e;
+                } catch (Throwable e) {
+                  throw new RuntimeException(e);
+                }
+              };
+          if (ref.isDefined()) {
+            // wrap the call if ContextRef contains a set Context
+            call = ref.get().wrap(call);
+          }
+          try {
+            return call.call();
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        },
+        ret);
+  }
+
+  public static <C extends Channel> ChannelInitializer<C> wrapServer(ChannelInitializer<C> inner) {
```

Comment:
ideally this should be handled by the netty instrumentation? is this correct?

Reply:
This is a strong maybe. In an ideal world, yes, that would be nice. But I'm not sure how that generally would apply, at least in this case bc of how finagle uses netty under the hood: they wire up their own ChannelInitializer [here](https://github.com/twitter/finagle/blob/d905fa7995f18ba8573d7e8f9983ba939b1c57d3/finagle-http2/src/main/scala/com/twitter/finagle/http2/transport/common/H2StreamChannelInit.scala#L32-L61) so this instrumentation exists to make sure that the finagle H2Initializer is appropriately instrumented.

In this case, I know this is where the instrumentation needs to happen bc of how all the finagle-netty bits are added, removed, modified, etc. But I don't know how or if this ChannelInitializer would apply to all cases more generally.

As for the static instrumentation method contents themselves ([this bit](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10141/files/45ec47cf495316e4c99a0354cd0fb6dedf8a683e#diff-c3a4a368b4d6c88d16614dc799c7cf04d93656c16f594a19d8dcacf312f9d881R160-R173), anyway), oh yes, that could probably just be extracted and exposed -- possibly in one of the netty library modules? I mention that specific block only bc in this case I'm specifically seeking out `Http2StreamFrameToHttpObjectCodec` which may not be the same for all use cases afaiui.

Want me to peel out that bit into netty while I'm at this?